### PR TITLE
Implement BAM format inputs

### PR DESCRIPTION
**Pull request by kshitij3579**
Added a bam2fastq function. BAM inputs should work and backbone filtering was done using this function.

### CODE REVIEW

File: FabFos.py [hunk 1]
```diff
@@ -215,6 +215,17 @@ def format(self, record):
 
 #################################### Classes end ################################################
 
+def bam2fastq(input_file, sample_id, output_dir, executables):
+    fastq_extract = [executables["bam2fastq"], "-o",
+                     output_dir + os.sep + sample_id + input_file + ".fasta", "--no-aligned", input_file]
+    fastq_extract += ["1>", output_dir + os.sep + "bam2fastq.stdout"]
+    fastq_extract += ["2>", output_dir + os.sep + "bam2fastq.stderr"]
+    p_bam2fastq = subprocess.Popen(' '.join(fastq_extract), shell=True, preexec_fn=os.setsid)
```
Comment (cmorganl) [inline FabFos.py]: Setting the `shell` parameter to `True` is a security issue and though this probably won't be critical to FabFos, it is frowned upon. Instead, I suggest removing the `' '.join()` call and setting `shell` to False.

File: FabFos.py [hunk 2]
```diff
@@ -215,6 +215,17 @@ def format(self, record):
 
 #################################### Classes end ################################################
 
+def bam2fastq(input_file, sample_id, output_dir, executables):
+    fastq_extract = [executables["bam2fastq"], "-o",
+                     output_dir + os.sep + sample_id + input_file + ".fasta", "--no-aligned", input_file]
```
Comment (cmorganl) [inline FabFos.py]: Typo in output path name - these should be ".fastq" not ".fasta". Also, let's use `_#` in the path in the same string as ".fastq" - `bam2fsatq` uses this to specify the forward and reverse reads.

File: FabFos.py [hunk 3]
```diff
@@ -228,17 +239,16 @@ def get_dict_keys(dictionary):
 
 def get_options():
     parser = argparse.ArgumentParser(description="Pipeline for processing and organizing fosmid sequence information.\n"
-                                                 "NOTE: a maximum of 2 sequence files are permitted.",
-                                     add_help=False)
+                                                 "NOTE: a maximum of 2 sequence files are permitted.",add_help=False)
```
Comment (cmorganl) [inline FabFos.py]: This can be changed back to the original version.

File: FabFos.py [hunk 4]
```diff
@@ -247,6 +257,9 @@ def get_options():
                           action="store_true", default=False, required=False)
 
     opts = parser.add_argument_group(title="Optional arguments")
+
+    opts.add_argument("-t", "--type", choices=["B", "F"], required=False, default="F",
+                      help="Enter B if input type is BAM, F for FastQ")
```
Comment (cmorganl) [inline FabFos.py]: Always a good idea to leave a note with what the default value is e.g. [ DEFAULT = 'F' ]

File: FabFos.py [hunk 5]
```diff
@@ -266,7 +279,9 @@ def get_options():
                       help="Increase the level of verbosity in runtime log.")
     opts.add_argument("-h", "--help",
                       action="help", help="Show this help message and exit")
-
+    if args.type == "B":
+        args.reads = bam2fastq(args.reads, sample.id, sample.output_dir, args.executables)
+
```
Comment (cmorganl) [inline FabFos.py]: This code should come outside of the `get_options` function. I suggest immediately after the call to `get_options` in the main function of FabFos.

**Review by cmorganl — CHANGES_REQUESTED** [review]
This branch will be merged to main once the requested changes have been made. Overall, looks like it should work and is a useful addition to the tool.

**Comment by kshitij3579** [comment]
I've made all the requested changes. Should I now submit another pull request?

**Review by cmorganl — APPROVED** [review]

**Pull request merged**